src/nvim/eval.c

```diff
@@ -13416,14 +13416,12 @@ static void f_resolve(typval_T *argvars, typval_T *rettv, FunPtr fptr)
           q = (char *)path_tail((char_u *)p);
         }
         if (q > p && !path_is_absolute_path((const char_u *)buf)) {
-          // Symlink is relative to directory of argument.
+          // Symlink is relative to directory of argument. Replace the
+          // symlink with the resolved name in the same directory.
           const size_t p_len = strlen(p);
           const size_t buf_len = strlen(buf);
-          cpy = xmalloc(p_len + buf_len + 1);
-          memcpy(cpy, p, p_len);
```

Comment:
Yes. I was thinking it's too easy to make mistakes like this. Especially when doing larger refactoring.

We may be able to catch a couple by running ASAN tests with a fully loaded (plugins) Vim doing all sorts of things (i.e.: what happened to me, though I didn't see it on OSX because it perhaps returns zeroed memory). However I think in general we should defer to using safer string functions. If `path_tail` didn't return `char_u` and take `char_u`, I would've used `xstrlcpy()`. Performance wise, this function does worse things than calculating the length of the string (in cache) redundantly. (Moreover, if we truly want that, `strlcpy` can be optimized a lot, copying most of glibcs `str(n)cpy`).